app/controllers/v1/terms_controller.rb

```diff
@@ -95,6 +99,15 @@ def options
     end
 
     private
+      def term_found?
+        if params[:term_type] == Term::TEMPORARY
+          Term.find_by(vocabulary: vocabulary, pref_label: params[:pref_label], term_type: Term::TEMPORARY)
```

Comment:
We don't want to do a find_by here with `pref_label` because it's not indexed on the db-side.  That could be an expensive lookup once there are lots of terms.  Same goes for `term_type`.  But actually, since this `term_found?` method is only used during create, I don't think you need it at all.  We can rely on a db unique constraint violation to catch this (index on `[:uri_hash, :vocabulary_id]`).

Reply:
ohhhh okay. That's a really good point. Do you think that instead of doing a begin/rescue because those are also expensive we could do something like the follow, that ways the error is caught during validation.
```
validates :uri_hash, uniqueness: { scope: :vocabulary }
```

Reply:
Ahh, I just realized I can't really do that and provide the 409. hmmm.

Reply:
The rescue expense might be partially negated by the savings we get from not doing an extra DB query to check if the term exists.  Also, failed create operations will probably be relatively rare, so the expense isn't too bad.

Is it expensive to wrap in begin/rescue, or is it only expensive if the rescue is invoked?

Reply:
If you put `render json: URIService::JSON.errors('Term already exists.'), status: 409` in the rescue and wrap the rest of the method's code in begin/rescue, doesn't that work for still returning 409?

Reply:
Yes, and no. It would work if we remove the model-level validations for uniqueness, which i don't want to do. I have an idea tho. How does this look:
```
    def create
      term = Term.new(create_params)
      term.vocabulary = vocabulary

      custom_fields.each do |f, v|
        next unless params.key?(f)
        term.set_custom_field(f, params[f])
      end

      if term.save
        render json: URIService::JSON.term(term), status: 201
      elsif term.errors.added?(:uri, :taken)
        render json: URIService::JSON.errors('Term already exists.'), status: 409
      else
        render json: URIService::JSON.errors(term.errors.full_messages), status: 400
      end
    end
```

I just tried it on my local instance and it works. I think this is neater than trying to catch errors.

Reply:
Relying on the uniqueness validation (as opposed to db constraint) works most of the time, and I wouldn't want to get rid of it either, but technically a db unique constraint violation can still occur:

```
2.11 uniqueness
This helper validates that the attribute's value is unique right before the object gets saved. It does not create a uniqueness constraint in the database, so it may happen that two different database connections create two records with the same value for a column that you intend to be unique. To avoid that, you must create a unique index on that column in your database.
```
(https://guides.rubyonrails.org/active_record_validations.html#uniqueness)

I guess in that case, the controller-wide rescue we talked about would still return an error to the client.  If `term.errors.added?(:uri, :taken)` returns true, doesn't that already give us an error on the object that notifies the client that the term already exists (so we don't need to handle that in a special way here)?

Reply:
I guess what I'm saying is, if we can rely on the controller to handle and display the unique constraint violation exception when it comes up, you can remove:

```
elsif term.errors.added?(:uri, :taken)
        render json: URIService::JSON.errors('Term already exists.'), status: 409
```

And then I think everything should work as intended.

Reply:
In order for the error to bubble up we would have to remove the model level validation, which we agreed we don't want to do. Adding the check for the validation error will catch the problem most of the time and will provide a helpful error message. Like any validation, if there's a problem when the db save happens then that would be a error that gets bubbled up.

Reply:
What I meant was what that the activemodel errors that are already present on the object (:uri and :taken) are already an indication that the term exists, right?  I don't know what we're adding by intercepting those errors and returning 'Term already exists.' instead of just showing those errors in the `else` section logic that you proposed.

But on an unrelated note, here's another problem: that uniqueness validation on the `:uri` field does an SQL lookup that selects `WHERE uri = x`, and uri is not an indexed field.  That's going to get really slow down the line.  So I think we actually must rely on a unique constraint resque in this case.  The usual solution to this problem would be to add a db index on the uri column, but we didn't do that in the first place because it's too long.  That's why we added that uri_hash/vocabulary_id field unique index.

Reply:
Yes, we can totally change the uniqueness check to check the uri_hash instead of uri. I was going to add it as an additional check, but I can just replace it. That's a good idea.

I also agree with just changing the status instead of returning a different error message.

I think we're on the same page, so I'll got ahead and make those changes!